core/vm/analysis.go

```diff
@@ -16,6 +16,102 @@
 
 package vm
 
+import (
+	"github.com/holiman/uint256"
+
+	"github.com/ledgerwatch/turbo-geth/common"
+)
+
+// fill in segment of operation information array for a block
+func analyzeBlock(ctx *callCtx, pc uint64) (*BlockInfo, error) {
+	blockInfo := NewBlockInfo(pc)
+	ctx.contract.opsInfo[pc] = blockInfo
+	code := ctx.contract.Code
+	codeLen := len(code)
+	jumpTable := ctx.interpreter.jt
+
+	height := 0
+	minHeight := 0
+	maxHeight := 0
+	for ; pc < uint64(codeLen); pc++ {
+		op := OpCode(code[pc])
+		oper := jumpTable[op]
+		if oper == nil {
+			continue
+		}
+
+		// track low and high watermark relative to block entry
+		height -= oper.numPop
+		minHeight = min(minHeight, height)	// will be <= 0
+		height += oper.numPush
+		maxHeight = max(maxHeight, height)	// will be >= 0
+		blockInfo.constantGas += oper.constantGas
+
+		if PUSH1 <= op && op <= PUSH32 {
+		    pushByteSize := int(op) - int(PUSH1) + 1
+
+			startMin := int(pc + 1)
+			if startMin >= codeLen {
+				startMin = codeLen
+			}
+			endMin := startMin + pushByteSize
+			if startMin+pushByteSize >= codeLen {
+				endMin = codeLen
+			}
+
+			integer := new(uint256.Int)
+			integer.SetBytes(common.RightPadBytes(
+				// So it doesn't matter what we push onto the stack.
+				code[startMin:endMin], pushByteSize))
+
+			// attach PushInfo with decoded push data to PUSHn
+			pushInfo := NewPushInfo(pc, *integer)
+			ctx.contract.opsInfo[pc] = pushInfo
+
+			continue
+		}
+		
+		// check jump destinations and optimize static jumps
+		if op == JUMP || op == JUMPI {
+			prevPC := pc - 1
+			prevOp := OpCode(code[prevPC])
+			if prevOp >= PUSH1 && prevOp <= PUSH32 {
+
+				pos := ctx.contract.opsInfo[prevPC].(PushInfo).data
+				if valid, _ := ctx.contract.validJumpdest(&pos); !valid {
+					return nil, ErrInvalidJump
+				}
+				jumpInfo := NewJumpInfo(pc, pos.Uint64())
+				ctx.contract.opsInfo[pc] = jumpInfo
+				ctx.contract.opsInfo[prevPC] = nil
+
+				// replace with JMP NOOP or JMPI NOOP and attach JumpInfo to JMP or JMPI
+				if op == JUMP {
+					code[prevPC] = byte(JMP)
+				}
+				if op == JUMPI {
+					code[prevPC] = byte(JMPI)
+				}
+				code[pc] = byte(NOOP)
```

Comment:
Looks dangerous. Are you sure the `code` slice is copied? Otherwise, this modifies the `state_object` code slice, and potentially a codesice living in the code cache.

Reply:
No, I'm not sure it's copied, but indeed it needs to be.

Reply:
@holiman  I'm copying Contract.Code to callCtx.code and modifying and interpreting only the copy.  I think that should do the trick.  Also appending 32 null  bytes to the copy, which might allow for simpler pops and pushes at the end of the code.